### PR TITLE
Add some property typehints

### PR DESCRIPTION
**Pull request by franmomu**
|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | no
| Fixed issues | 

#### Summary

This add some property type hints to `tests` directory,  I've changed a couple of things that I would point out in the code. Since there are a lot missing, I'll try to make manageable PRs.

### CODE REVIEW

File: tests/Doctrine/ODM/MongoDB/Tests/Functional/Ticket/GH1294Test.php [hunk 1]
```diff
@@ -43,8 +43,12 @@ class GH1294User
...
     public $id;
 
-    /** @ODM\Field(type="string") */
-    public $name = false;
+    /**
+     * @ODM\Field(type="string")
+     *
+     * @var string|null
+     */
+    public $name = '';
```
Comment (franmomu) [inline tests/Doctrine/ODM/MongoDB/Tests/Functional/Ticket/GH1294Test.php:51]: `name` is always initialized:

https://github.com/doctrine/mongodb-odm/blob/15773364718445ce999cdf501d297b5be31e25af/tests/Doctrine/ODM/MongoDB/Tests/Functional/Ticket/GH1294Test.php#L15-L21

File: tests/Doctrine/ODM/MongoDB/Tests/Persisters/PersistenceBuilderTest.php [hunk 2]
```diff
@@ -76,7 +76,7 @@ public function testFindWithOrOnCollectionWithDiscriminatorMap(): void
 
         $sameCollection1->id   = $ids[0];
         $sameCollection1->name = 'First entry in SameCollection1';
-        $sameCollection1->test = 1;
+        $sameCollection1->test = 'test';
```
Comment (franmomu) [inline tests/Doctrine/ODM/MongoDB/Tests/Persisters/PersistenceBuilderTest.php:79]: It is used as `string`:

https://github.com/doctrine/mongodb-odm/blob/15773364718445ce999cdf501d297b5be31e25af/tests/Doctrine/ODM/MongoDB/Tests/Persisters/PersistenceBuilderTest.php#L55-L63

File: lib/Doctrine/ODM/MongoDB/Configuration.php [hunk 3]
```diff
@@ -402,11 +406,15 @@ public function getDefaultDB(): ?string
...
     }
 
+    /** @psalm-param class-string $cmfName */
     public function setClassMetadataFactoryName(string $cmfName): void
     {
         $this->attributes['classMetadataFactoryName'] = $cmfName;
     }
 
+    /**
+     * @psalm-return class-string
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Configuration.php]: Let's be consistent with multi-line vs single-lined PHPDocs :)

**Review by malarzm — COMMENTED** [review]
I'm gonna trust you and all our tests with this :)

**Review by malarzm — APPROVED** [review]

File: lib/Doctrine/ODM/MongoDB/Configuration.php [hunk 4]
```diff
@@ -87,8 +87,11 @@ class Configuration
      * @psalm-var array{
      *      autoGenerateHydratorClasses?: self::AUTOGENERATE_*,
      *      autoGeneratePersistentCollectionClasses?: self::AUTOGENERATE_*,
+     *      classMetadataFactoryName?: class-string,
```
Comment (IonBazan) [inline lib/Doctrine/ODM/MongoDB/Configuration.php]: Let's try to limit allowed classes here. It appears that it should only accept `\Doctrine\ODM\MongoDB\Mapping\ClassMetadataFactory`. In fact, the class is marked as `final` so I can't imagine anyone extending it but you can still create your own implementation that adds `setDocumentManager` and `getDocumentManager` methods (there is no common interface for that).

File: lib/Doctrine/ODM/MongoDB/Configuration.php [hunk 5]
```diff
@@ -87,8 +87,11 @@ class Configuration
...
      *      defaultCommitOptions?: CommitOptions,
      *      defaultDocumentRepositoryClassName?: string,
+     *      defaultGridFSRepositoryClassName?: class-string,
```
Comment (IonBazan) [inline lib/Doctrine/ODM/MongoDB/Configuration.php]: This should implement `\Doctrine\ODM\MongoDB\Repository\GridFSRepository`

File: tests/Doctrine/ODM/MongoDB/Tests/Events/LifecycleListenersTest.php [hunk 6]
```diff
@@ -68,6 +72,7 @@ public function testLifecycleListeners(): void
         $this->listener->called = [];
 
         $document = $dm->find(TestDocument::class, $test->id);
+        assert($document->embedded instanceof PersistentCollectionInterface);
```
Comment (IonBazan) [inline tests/Doctrine/ODM/MongoDB/Tests/Events/LifecycleListenersTest.php]: Can we use `self::assertInstanceOf()` together with `phpstan-phpunit` plugin instead?

File: lib/Doctrine/ODM/MongoDB/Configuration.php (hunk 3 above)
Comment (franmomu) [reply lib/Doctrine/ODM/MongoDB/Configuration.php]: done, I've used multiline for PHPDoc in methods (I've found a way to force this from phpcs), the single-lined one is forced by phpcs when only has one single line content.

File: lib/Doctrine/ODM/MongoDB/Configuration.php (hunk 4 above)
Comment (franmomu) [reply lib/Doctrine/ODM/MongoDB/Configuration.php]: sure, I've used `class-string<ClassMetadataFactory>` and modified also `defaultGridFSRepositoryClassName` and `classMetadataFactoryName`, thanks!

File: tests/Doctrine/ODM/MongoDB/Tests/Events/LifecycleListenersTest.php (hunk 6 above)
Comment (franmomu) [reply tests/Doctrine/ODM/MongoDB/Tests/Events/LifecycleListenersTest.php]: yep, done.

File: lib/Doctrine/ODM/MongoDB/Configuration.php (hunk 4 above)
Comment (IonBazan) [reply lib/Doctrine/ODM/MongoDB/Configuration.php]: Can you also apply that to the getters and setters where applicable?

**Review by IonBazan — APPROVED** [review]

File: lib/Doctrine/ODM/MongoDB/Configuration.php (hunk 4 above)
Comment (franmomu) [reply lib/Doctrine/ODM/MongoDB/Configuration.php]: yeah, should be better now, thanks.

**Pull request merged**